qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -0,0 +1,175 @@
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, Optional
+
+from ..problems.quadratic_program import QuadraticProgram
+
+
+class OptimizationAlgorithm(ABC):
+    """An abstract class for optimization algorithms in Qiskit Optimization."""
+
+    @abstractmethod
+    def get_compatibility_msg(self, problem: QuadraticProgram) -> str:
+        """Checks whether a given problem can be solved with the optimizer implementing this method.
+
+        Args:
+            problem: The optimization problem to check compatibility.
+
+        Returns:
+            Returns the incompatibility message. If the message is empty no issues were found.
+        """
+
+    def is_compatible(self, problem: QuadraticProgram) -> bool:
+        """Checks whether a given problem can be solved with the optimizer implementing this method.
+
+        Args:
+            problem: The optimization problem to check compatibility.
+
+        Returns:
+            Returns True if the problem is compatible, False otherwise.
+        """
+        return len(self.get_compatibility_msg(problem)) == 0
+
+    @abstractmethod
+    def solve(self, problem: QuadraticProgram) -> 'OptimizationResult':
+        """Tries to solves the given problem using the optimizer.
+
+        Runs the optimizer to try to solve the optimization problem.
+
+        Args:
+            problem: The problem to be solved.
+
+        Returns:
+            The result of the optimizer applied to the problem.
+
+        Raises:
+            QiskitOptimizationError: If the problem is incompatible with the optimizer.
+        """
+        raise NotImplementedError
+
+
+class OptimizationResultStatus(Enum):
+    """Feasible values for the termination status of an optimization algorithm.
+    """
+    SUCCESS = 0
+    FAILURE = 1
+    INFEASIBLE = 2
+
+
+class OptimizationResult:
+    """The optimization result class.
+
+    The optimization algorithms return an object of the type `OptimizationResult`, which enforces
+    providing the following attributes.
+
+    Attributes:
+        x: The optimal value found in the optimization algorithm.
+        fval: The function value corresponding to the optimal value.
+        results: The original results object returned from the optimization algorithm. This can
+            contain more information than only the optimal value and function value.
+        status: The termination status of the algorithm.
+    """
+
+    Status = OptimizationResultStatus
+
+    def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
+                 results: Optional[Any] = None,
+                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS) -> None:
+        """Initialize the optimization result."""
+        self._val = x
+        self._fval = fval
+        self._results = results
+        self._status = status
+
+    def __repr__(self):
+        return '([%s] / %s)' % (','.join([str(x_) for x_ in self.x]), self.fval)
```

Comment:
I the notebook output when this is printed we get `([0.0,1.0,0.0] / -2.0)` for instance. Maybe this is a suitable __repr__ and maybe it might help to have a more user friendly __str__ for printing? With the surround () and the / it looks more like a single bracketed formula than 2 items. Would something like vars=[0.0,1.0,0.0], min=-2.0 do which seems to be more easily parsed/consumed by an end user